containerize/command.go

```diff
@@ -0,0 +1,261 @@
+// Sets up Cgroup-based limits for the process running the command
+func (c *Command) setupLimits() error {
+	var err error
+
+	// Create a cgroup for this command
+	c.cgroup, err = cgroups.NewContainer(c.CommandId)
+	if err != nil {
+		return fmt.Errorf("failed to setup cgroups for command %s: %w", c.CommandId, err)
+	}
+
+	// Add the process to the cgroup
+	c.cgroup.AddProcess(c.pid())
+
+	// Set up limits
+	if err := c.cgroup.MemoryLimitBytes(10 * 1024 * 1024); err != nil { // 10Mb limit
+		return fmt.Errorf("failed to set a memory limit: %w", err)
+	}
+	if err := c.cgroup.IoWeight(1); err != nil { // Put the lowest IO weight on this command
+		return fmt.Errorf("failed to set a io weight limit: %w", err)
+	}
+	if err := c.cgroup.CpuLimitPct(10); err != nil { // 10% of CPU maximum per command
+		return fmt.Errorf("failed to set a CPU limit: %w", err)
+	}
+	return nil
+}
+
+//-------------------------------------------------------------------------------------------------
+func (c *Command) sysProcAttr() *syscall.SysProcAttr {
+	attr := syscall.SysProcAttr{
+		// Use the same process group for all processed spawned by the command
+		// This is needed to make sure we can kill the whole tree at once
+		Setpgid: true,
+
+		// Make sure all processes die when the server is killed
+		Pdeathsig: syscall.SIGKILL,
+
+		// Isolate the process as needed
+		Cloneflags: syscall.CLONE_NEWUTS | // New UTS IPC namespace (isolated hostname, etc)
+			syscall.CLONE_NEWPID | // Isolated PID namespace
+			syscall.CLONE_NEWNET, // Isolated network environment
+
+		// This requires more magic if we want to really chroot somewhere (by itself it does not do much)
+		// syscall.CLONE_NEWNS | // Isolated mount namespace
+
+		// syscall.CLONE_NEWUSER | // Isolated user namespace
+		// Whatever uid/gid we use for the server will be mapped into root within the container
+		// UidMappings: []syscall.SysProcIDMap{
+		// 	{
+		// 		ContainerID: 0,
+		// 		HostID:      os.Getuid(),
+		// 		Size:        1,
+		// 	},
+		// },
+		// GidMappings: []syscall.SysProcIDMap{
+		// 	{
+		// 		ContainerID: 0,
+		// 		HostID:      os.Getgid(),
+		// 		Size:        1,
+		// 	},
+		// },
+	}
+
+	return &attr
+}
+
+//-------------------------------------------------------------------------------------------------
+// Returns true if the command is still running
+func (c *Command) Running() bool {
+	c.commandMutex.RLock()
+	running := c.running
+	c.commandMutex.RUnlock()
+	return running
+}
+
+// Waits for the process to finish (without relying on waitpid, which destroys process exit status info)
+func (c *Command) Wait() {
+	for c.Running() {
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+//-------------------------------------------------------------------------------------------------
+// Terminates the command if it is running (including all sub-processes)
+func (c *Command) Kill() {
+	if c.Running() {
```

Comment:
This looks racy. You're taking out two separate and independent locks on the mutex.

`c.running` could be true on the initial check, and get set to false before you re-acquire the lock.

Reply:
Updated to direct reading from `c.running` while under a lock